File: java/SavingsAccount.java
package com.bank;

/**
 * SavingsAccount demonstrating Inheritance and Polymorphism.
 * Enforces a Minimum Balance requirement (₹1,000) and provides 4% annual interest.
 */
public class SavingsAccount extends BankAccount {
    public static final double MINIMUM_BALANCE = 1000.0;
    public static final double ANNUAL_INTEREST_RATE = 4.0;
    public static final double DEFAULT_DAILY_LIMIT = 50000.0;

    public SavingsAccount(String accountNumber, Customer customer, double initialDeposit) {
        super(accountNumber, customer, initialDeposit, DEFAULT_DAILY_LIMIT);
        if (initialDeposit < MINIMUM_BALANCE) {
            System.out.println("Warning: Savings account initial deposit is below the minimum balance of ₹" + MINIMUM_BALANCE);
        }
    }

    public SavingsAccount(String accountNumber, Customer customer, double initialDeposit, double dailyLimit) {
        super(accountNumber, customer, initialDeposit, dailyLimit);
    }

    @Override
    public String getAccountType() {
        return "Savings Account";
    }

    @Override
    public double getMinimumBalance() {
        return MINIMUM_BALANCE;
    }

    @Override
    public double getOverdraftLimit() {
        return 0.0; // Savings accounts do not permit overdraft
    }

    @Override
    public double getInterestRate() {
        return ANNUAL_INTEREST_RATE;
    }

    /**
     * Polymorphic implementation: Ensures balance does not drop below Minimum Balance (₹1,000).
     */
    @Override
    protected boolean canWithdraw(double amount, StringBuilder reason) {
        if ((balance - amount) < MINIMUM_BALANCE) {
            reason.append("Insufficient funds. Maintaining minimum balance of ₹")
                  .append(MINIMUM_BALANCE)
                  .append(" is required. Current balance: ₹")
                  .append(balance);
            return false;
        }
        return true;
    }

    /**
     * Applies monthly/quarterly interest based on annual interest rate.
     */
    public void applyInterest() {
        double interest = (balance * (ANNUAL_INTEREST_RATE / 100.0)) / 12.0;
        if (interest > 0) {
            deposit(interest, "Monthly Interest Credit @ " + ANNUAL_INTEREST_RATE + "% p.a.");
        }
    }
}

File: src/models/SavingsAccount.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { BankAccount } from './BankAccount';
import { Customer } from './Customer';
import { AccountType, OperationResult } from './types';
import { Transaction } from './Transaction';

/**
 * Savings Account subclass.
 * Demonstrates:
 * - Inheritance: Extends BankAccount
 * - Polymorphism: Custom withdrawal validation enforcing Minimum Balance, Interest computation
 */
export class SavingsAccount extends BankAccount {
  private static readonly MINIMUM_BALANCE = 1000.0; // ₹1,000 minimum required balance
  private static readonly ANNUAL_INTEREST_RATE = 4.0; // 4.0% annual interest
  private static readonly DEFAULT_DAILY_LIMIT = 25000.0; // ₹25,000 default daily withdrawal limit

  constructor(
    accountNumber: string,
    customer: Customer,
    initialDeposit: number,
    customDailyLimit?: number,
    openedDate: Date = new Date()
  ) {
    if (initialDeposit < SavingsAccount.MINIMUM_BALANCE) {
      throw new Error(
        `Savings Account requires an initial deposit of at least ₹${SavingsAccount.MINIMUM_BALANCE}. Received: ₹${initialDeposit}`
      );
    }

    super(
      accountNumber,
      customer,
      initialDeposit,
      customDailyLimit || SavingsAccount.DEFAULT_DAILY_LIMIT,
      openedDate
    );
  }

  public getAccountType(): AccountType {
    return AccountType.SAVINGS;
  }

  public getMinimumBalance(): number {
    return SavingsAccount.MINIMUM_BALANCE;
  }

  public getOverdraftLimit(): number {
    return 0; // Savings accounts do not permit overdraft
  }

  public getInterestRate(): number {
    return SavingsAccount.ANNUAL_INTEREST_RATE;
  }

  /**
   * Polymorphic fund sufficiency check for Savings Account:
   * Must preserve the $100 minimum balance buffer.
   */
  protected canWithdrawFunds(amount: number): { allowed: boolean; reason?: string } {
    const currentBalance = this.getBalance();
    const remainingAfterWithdrawal = currentBalance - amount;

    if (remainingAfterWithdrawal < SavingsAccount.MINIMUM_BALANCE) {
      return {
        allowed: false,
        reason: `Savings Account requires maintaining a minimum balance of ₹${SavingsAccount.MINIMUM_BALANCE.toFixed(
          2
        )}. Current balance: ₹${currentBalance.toFixed(2)}, requested: ₹${amount.toFixed(
          2
        )}. Max allowable withdrawal: ₹${Math.max(0, currentBalance - SavingsAccount.MINIMUM_BALANCE).toFixed(2)}.`,
      };
    }

    return { allowed: true };
  }

  /**
   * Polymorphic periodic interest application for Savings Account.
   * Credits monthly interest (AnnualRate / 12).
   */
  public applyPeriodicInterest(date: Date = new Date()): OperationResult<Transaction> {
    const balance = this.getBalance();
    if (balance <= 0) {
      return {
        success: false,
        message: 'No interest applicable for zero or negative balance.',
        timestamp: new Date(),
      };
    }

    const monthlyRate = (SavingsAccount.ANNUAL_INTEREST_RATE / 100) / 12;
    const interestAmount = Math.round(balance * monthlyRate * 100) / 100;

    if (interestAmount <= 0) {
      return {
        success: false,
        message: 'Accrued interest is less than 1 cent.',
        timestamp: new Date(),
      };
    }

    const txn = this.recordSystemCredit(
      interestAmount,
      `Monthly Interest Credit (${SavingsAccount.ANNUAL_INTEREST_RATE}% p.a.)`,
      date
    );

    return {
      success: true,
      message: `Credited $${interestAmount.toFixed(2)} interest into Savings Account.`,
      data: txn,
      timestamp: new Date(),
    };
  }
}
